rasterio/_err.pyx

```diff
@@ -219,35 +230,47 @@ cdef OGRErr exc_wrap_ogrerr(OGRErr err) except -1:
         raise CPLE_BaseError(3, err, "OGR Error code {}".format(err))
 
 
+cdef class StackChecker:
+
+    def __init__(self, error_stack=None):
+        self.error_stack = error_stack or {}
+
+    cdef int exc_wrap_int(self, int err) except -1:
+        """Wrap a GDAL/OGR function that returns CPLErr (int).
+
+        Raises a Rasterio exception if a non-fatal error has be set.
+        """
+        if err:
+            stack = self.error_stack.get()
+            for error, cause in zip_longest(stack[::-1], stack[::-1][1:]):
+                if error is not None and cause is not None:
+                    error.__cause__ = cause
+
+            if stack:
+                last = stack.pop()
+                if last is not None:
+                    raise last
+
+        return err
+
+
 @contextlib.contextmanager
 def stack_errors():
-    """TODO: better name."""
-    # Set up.
+    # TODO: better name?
+    # Note: this manager produces one chain of errors and thus assumes
+    # that no more than one GDAL function is called.
     CPLErrorReset()
-    error_stack = []
 
     # chaining_error_handler (better name a TODO) records GDAL errors
     # in the order they occur and converts to exceptions.
-    CPLPushErrorHandlerEx(<CPLErrorHandler>chaining_error_handler, <void *>error_stack)
+    CPLPushErrorHandlerEx(<CPLErrorHandler>chaining_error_handler, NULL)
```

Comment:
The handler looks up the thread local chained error stack, no need to pass a pointer to data.